web/packages/teleterm/src/ui/services/workspacesService/documentsService/types.ts

```diff
@@ -35,18 +35,39 @@ export interface DocumentBlank extends DocumentBase {
   kind: 'doc.blank';
 }
 
-export interface DocumentTshNode extends DocumentBase {
+export type DocumentTshNode =
+  | DocumentTshNodeWithServerId
+  | DocumentTshNodeWithLoginHost;
+
+interface DocumentTshNodeBase extends DocumentBase {
   kind: 'doc.terminal_tsh_node';
   status: '' | 'connecting' | 'connected' | 'disconnected';
+  rootClusterId: string;
+  leafClusterId: string | undefined;
+}
+
+export interface DocumentTshNodeWithServerId extends DocumentTshNodeBase {
+  // serverId is the UUID of the SSH server. If it's is present, we can immediately start an SSH
+  // session.
+  //
+  // serverId is available when connecting to a server from the resource table.
   serverId: string;
+  // serverUri is used for file transfer and for identifying a specific server among different
+  // profiles and clusters.
   serverUri: uri.ServerUri;
-  rootClusterId: string;
-  leafClusterId?: string;
   // login is missing when the user executes `tsh ssh user` from the command bar without supplying
   // the hostname. In that case, login will be undefined and serverId will be equal to "user".
   //
   // We will execute `tsh ssh user` anyway and let tsh show an appropriate error message.
   login?: string;
+  loginHost?: string;
```

Comment:
I have trouble understanding it. Why this document has `loginHost`? Don't we want to have it only in `DocumentTshNodeWithLoginHost`?
>In short, I take the existing `DocumentTshNode` type and split it into two separate types. One has the `loginHost` field and the other one has `serverId`, `serverUri` and `login`.

Reply:
> I have trouble understanding it. Why this document has `loginHost`? Don't we want to have it only in `DocumentTshNodeWithLoginHost`?

`DocumentsService.update` doesn't really allow us to _remove_ fields from an object so I figured I might keep the type truthful to the implementation (which is something we should avoid doing, really 😶‍🌫️; it should work the other way around.).

I could change it to `loginHost?: undefined` since we overwrite this field anyway AFAIR.

Reply:
I pushed a commit to the next PR which makes it undefined.

Reply:
`DocumentsService.update` is an odd one anyway. I was considering renaming it to `unsoundUpdate`/`unsafeUpdate` and adding a generic to `update` so that it lets you change only the fields of the doc you're passing in. Ultimately I scratched that idea because I didn't want to spend even more time on this PR.

Reply:
I like the idea of `unsafeUpdate`, but I agree, it is not worth to spend more time on this.
Could you add a comment there with a little explanation why we have this field?

Reply:
> Could you add a comment there with a little explanation why we have this field?

Added in the next PR.